pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,338 @@
+	LabelsMap    map[types.NamespacedName]*prometheus.Labels
+	GaugeVecMap  map[string]*prometheus.GaugeVec
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client, coreV1Client corev1client.CoreV1Interface) *Controller {
+	newcontroller := Controller{
+		KubeClient:   kubeClient,
+		CoreV1Client: coreV1Client,
+		LabelsMap:    make(map[types.NamespacedName]*prometheus.Labels),
+		GaugeVecMap:  make(map[string]*prometheus.GaugeVec),
+	}
+	numPodGaugeVec := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "numberofpods",
+			Help:      "Auto-generated gauge: numberofpods",
+		},
+		getLabelNames(),
+	)
+	newcontroller.GaugeVecMap["numberofpods"] = numPodGaugeVec
+	crmetrics.Registry.MustRegister(numPodGaugeVec)
+
+	overheadGaugeVec := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "overhead",
+			Help:      "Auto-generated gauge: overhead",
+		},
+		getLabelNames(),
+	)
+	newcontroller.GaugeVecMap["overhead"] = overheadGaugeVec
+	crmetrics.Registry.MustRegister(overheadGaugeVec)
+
+	return &newcontroller
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("nodemetrics").With("node", req.Name))
+	// Retrieve node from reconcile request
+	node := &v1.Node{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to node deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				c.deleteGaguges(labels)
+			} else {
+				logging.FromContext(ctx).Errorf("Failed to delete gauge: failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after node labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+		c.deleteGaguges(labels)
+	}
+	newlabels, err := c.generateLabels(node)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate labels: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	if err := c.updateGaguges(ctx, node, newlabels); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to update gauges: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+	c.LabelsMap[req.NamespacedName] = newlabels
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("nodemetrics").
+		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
```

Comment:
We should abstract this somewhere -- it's used multiple times. `MapProvisionerToNode`

Reply:
Lets change this in the next PR. I don't want to modify other controllers in this PR.